### ISSUE TITLE
persistentStoreCoordinator.migratePersistentStore() fails with SMStore.type 

### ISSUE THREAD

**Issue by tifroz** [paulw11/Seam3#45]
When migrating an existing sore (of type`NSSQLiteStoreType`) into a new one (of type `SMStore.type`), the Core Data stack throws an uncatchable error

```
2017-09-27 18:06:30.915044-0700 Seam3_Example[5110:986518] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSArrayM replaceObjectAtIndex:withObject:]: object cannot be nil'
*** First throw call stack:
(0x18201efe0 0x180a80538 0x181f01d2c 0x1843d58fc 0x1843e0630 0x100e49a10 0x100e565bc 0x1843cfd70 0x1843d52fc 0x1000ff424 0x1000ffce0 0x1881c1e48 0x1883ce37c 0x1883d3e24 0x1883e88b0 0x1883d10b8 0x183bc8884 0x183bc86f0 0x183bc8aa0 0x181fcd42c 0x181fccd9c 0x181fca9a8 0x181efada4 0x1881bafc8 0x1881b5c9c 0x100101ef8 0x180f0959c)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

The error can be duplicated by replacing the `didFinishLaunchingWithOptions` function in the *Example* project.

```
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
      
      let documentDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).last!
      let oldURL = documentDirectory.appendingPathComponent("old.sqlite")
      let newURL = documentDirectory.appendingPathComponent("new.sqlite")
      let modelUrl = Bundle.main.url(forResource: "Seam3Demo", withExtension: "momd")!
      if let mom = NSManagedObjectModel(contentsOf: modelUrl) {
        let container = NSPersistentContainer(name: "container", managedObjectModel: mom)
        
        let psc = container.persistentStoreCoordinator
        guard let store = try? psc.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: oldURL, options: nil) else {
          fatalError("Failed to create old store")
        }
        
        let device = NSEntityDescription.insertNewObject(forEntityName: "Device", into: container.viewContext) as! Device
        device.deviceID = "123"
        guard let saved = try? container.viewContext.save() else {
          fatalError("Can't save")
        }
        
        let exists = FileManager.default.fileExists(atPath: oldURL.path)
        guard let oldStore = psc.persistentStore(for: oldURL) else {
          fatalError("Failed to reference old store")
        }
        
        do {
          SMStore.registerStoreClass()
          SMStore.syncAutomatically = false
         // try psc.migratePersistentStore(oldStore, to: newURL, options:nil, withType: NSSQLiteStoreType) // Does not crash
          try psc.migratePersistentStore(oldStore, to: newURL, options:nil, withType:SMStore.type) // Crashes
          // We never make it here when using SMStore.type
        } catch {
          // ...or here
          fatalError("Failed to migrate store: \(error)")
        }
      }
}
```

I did a good deal of troubleshooting but have unfortunately little to show for it. Here is what I know:
- The single record created in the code above gets copied over to the new store as part of the migration, firing `SMStore`'s `func execute(_ request: NSPersistentStoreRequest, with context: NSManagedObjectContext?)`.
- The `execute` function ends normally, returning an empty array as it should with a request type `.saveRequestType`.
- The error occurs while executing core data stack code, before ` psc.migratePersistentStore()` actually completes.

In other words, there is something SMStore does while executing the insert command that upsets the Core Data stack ... but what could it be??

**Comment by paulw11** [comment]
This is interesting; it appears to be an Apple bug; they shouldn't do whatever they are doing with a `nil` object.  There are a few hits on this on the web, the fix in one case was source and target store urls being the same, and in the other case the guy just moved the database without the migration.

I set a breakpoint and I can see the record being inserted into the Seam store and I pulled the application container down and the migration has taken place; the new.sqllite contains the single device record.

It seems that there is some operation that occurs after the migration that is causing the crash.

I may have to create a smaller example and open a ticket with Apple.

**Comment by paulw11** [comment]
I have opened a case with Apple. I will let you know how I go.

**Comment by tifroz** [comment]
Thanks I really appreciate an update when you have one. I was planning to move my legacy ubiquity store to Seam and this bug is throwing a serious wrench into this plan :(

**Comment by DJ-Glock** [comment]
Hi guys.
Sorry for asking you here, but probably you can easily help me. I'm checking articles about lightweight migration, but what is the best practice to migrate existing SQL type database to Seam3?

For example I need to support adding new attributes for Seam3:

>     NSUnderlyingError = "Error Domain=NSCocoaErrorDomain Code=134110 \"(null)\" UserInfo={entity=GuestsTable, attribute=sm_LocalStore_RecordID, reason=Validation error missing attribute values on mandatory destination attribute}";
>     destinationURL = "file:///Users/dj-glock/Library/Developer/CoreSimulator/Devices/F17F412A-5DCB-40C2-B1CD-7BDDA8EE1EFB/data/Containers/Data/Application/33C4D94B-2C41-43A1-A027-B96FC0E5813E/Library/Application%20Support/.CafeManager.sqlite.migrationdestination_41b5a6b5c6e848c462a8480cd24caef3";
>     reason = "Cannot migrate store in-place: Validation error missing attribute values on mandatory destination attribute";
>     sourceURL = "file:///Users/dj-glock/Library/Developer/CoreSimulator/Devices/F17F412A-5DCB-40C2-B1CD-7BDDA8EE1EFB/data/Containers/Data/Application/33C4D94B-2C41-43A1-A027-B96FC0E5813E/Library/Application%20Support/CafeManager.sqlite";

May be you have some universal solutions that can be added to Seam3 documentation?

**Comment by tifroz** [comment]
@DJ-Glock If you take the time to read the thread above: sample code is provided + a critical piece of information which is that the built-in Core Data migration doesn't work with SMStore. You can either roll your own or wait for Apple to fix their code (if in fact a Core Data bug)

**Comment by DJ-Glock** [comment]
@tifroz Got it, thanks.

**Comment by DJ-Glock** [comment]
@paulw11 If it's CoreData bug (not swift), I guess it will be fixed only for iOS 11+, won't it?

**Comment by paulw11** [comment]
The crash does not occur with migration on other store types, so I am hopeful that Apple can tell me what Seam3 is doing that causes the crash and I can code around the issue. Judging by some similar crashes reported on StackOverflow this bug has been around for several years but isn’t commonly triggered. 

A fix where the process returns an error rather than crashing will need to be in a future release of iOS.

**Comment by paulw11** [comment]
Apple have informed me that the `NSStoreUUIDKey` needs to be a fixed value.  Can you try the `develop` branch and see if migration is working for you?  I have tested it and it seems good.

**Comment by tifroz** [comment]
Yes, it seems to be working for me as well.

I only performed a very basic test using the code I used to demonstrate the issue to begin with, but I can do more comprehensive testing after the PR for taking into account object graph dependencies is incorporated (I can't really do any meaningful migration until then)

**Comment by paulw11** [comment]
I have merged your PR and merged the new master into develop, so you should be able to test using the develop branch with both object graph insertion and the migration fix.

**Comment by tifroz** [comment]
My tests have all been successful using the dev branch so far (using moderately complex data)



On a separate note, I am planning a PR to add a new API to SMStore `verifyCloudKitStoreExists()`, because when migrating from a legacy ubiquity store, you must first find out if another device already performed the migration. The pseudo logic goes like this:
```
if (anotherDeviceDidMigrate) {
  deleteLegacyStore()
  createSMStoreDataStackAndSync()
} else {
  migrateFromLegacyToSMStoreAndSync()
  deleteLegacyStore()
}
```

The API should look something like this:
```
  open func verifyCloudKitStoreExists(_ completionHandler: ((_ exists: Bool, _ error: Error?) -> Void)?) {
    guard self.cloudKitValid else {
      NSLog("Access to CloudKit has not been verified by calling verifyCloudKitConnection")
      return
    }
    let operation = SMServerZoneLookupOperation(cloudDatabase: database)
    operation.lookupOperationCompletionBlock = completionHandler
    let queue = OperationQueue()
    queue.addOperation(operation)
  }
```

...with `SMServerZoneLookupOperation ` essentially testing if a zone named `SMStore.SMStoreCloudStoreCustomZoneName` does indeed exist.

Let me know if you have any suggestions / thoughts?

**Comment by DJ-Glock** [comment]
@tifroz Hey
Have you succeeded with implementing of logic when another device performed migration & sync?
I have some doubts. I have used this code for migration. Migration works. But after migration, Seam upload to iCloud only new changes, but not existing data, that was in old store.

```
    // MARK: - Core Data stack for iOS 10+
    @available(iOS 10.0, *)
    static var persistentContainer: NSPersistentContainer = {
        var applicationDocumentsDirectory: URL = {
            let urls = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
            return urls[urls.count-1]
        }()
        
        let modelURL = Bundle.main.url(forResource: "CafeManager", withExtension: "momd")!
        
        //Urls to support old database migration to Seam3 type
        let oldURL = applicationDocumentsDirectory.appendingPathComponent("CafeManager.sqlite")
        let newURL = applicationDocumentsDirectory.appendingPathComponent("CafeManagerSeam3.sqlite")
        print(oldURL)

        guard let managedObjectModel = NSManagedObjectModel(contentsOf: modelURL) else {fatalError("Failed to load managedObjectModel \(modelURL)")}
        let container = NSPersistentContainer(name: "container", managedObjectModel: managedObjectModel)
        
        let persistentStoreCoordinator = container.persistentStoreCoordinator
        guard let store = try? persistentStoreCoordinator.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: oldURL, options: nil) else {
            fatalError("Failed to create old store")
        }
        
        let exists = FileManager.default.fileExists(atPath: oldURL.path)
        guard let oldStore = persistentStoreCoordinator.persistentStore(for: oldURL) else {
            fatalError("Failed to reference old store")
        }
        
        do {
            SMStore.registerStoreClass()
            let newStore = try persistentStoreCoordinator.migratePersistentStore(oldStore, to: newURL, options: nil, withType:SMStore.type)
            let storeDescription = NSPersistentStoreDescription(url: newURL)
            
            storeDescription.type = SMStore.type
            storeDescription.setOption("iCloud.iGlock.CafeManager.com" as NSString, forKey: SMStore.SMStoreContainerOption)
            storeDescription.setOption(NSNumber(value:SMSyncConflictResolutionPolicy.clientTellsWhichWins.rawValue), forKey:SMStore.SMStoreSyncConflictResolutionPolicyOption)
            container.persistentStoreDescriptions=[storeDescription]
            
            try FileManager.default.removeItem(at: oldURL)
        } catch {
            fatalError("Failed to migrate store: \(error)")
        }
        
        return container
    }()
```

**Comment by tifroz** [comment]
@DJ-Glock I ran into this too. This is arguably a bug, which you can work around by creating the SMStore stack before you migrate to it. In other words, before you migrate, execute something like this:
```
try? persistentStoreCoordinator.addPersistentStore(ofType: SMStore.type, configurationName: nil, at: newURL, options: nil)
```

When you migrate before creating the `SMStore` first, the migration process does create the store for you, but the `SMStore` immediately detects a `new iCloud user` (you should see this in the logs), so it resets itself & deletes all the data created by the migration process.

Also, you should probably prevent automatic syncing before migration, and manually `triggerSync()` after the migration is complete. It's a lot cleaner, and a lot easier to troubleshoot if you migrate first, sync second as opposed to doing it all at the same time.

**Comment by tifroz** [comment]
Actually the bug is in the example app, not with the SMStore library itself (if that's what you are using as a baseline for testing). You should be able to fix the bug without changing any of the SMStore source files:

Make sure the store doesn't get reset on its first use by making sure `previousUser` is not nil before resetting the store in `validateCloudKitAndSync` (see below)

```
func validateCloudKitAndSync(_ completion:@escaping (() -> Void)) {
        self.smStore?.verifyCloudKitConnectionAndUser() { (status, user, error) in
            //...
            //if  previousUser != currentUser {   // REPLACE THIS LINE WITH LINE BELOW
            if  previousUser != currentUser && previousUser != nil {
                // ...
            }
            // ...
```
